File: app/src/main/java/com/nrs/nsnik/notes/view/MainActivity.java
/*
 * Copyright (C) 2017 nsnikhil
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.nrs.nsnik.notes.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.support.design.widget.NavigationView;
import android.support.test.espresso.IdlingResource;
import android.support.v4.app.Fragment;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.jakewharton.rxbinding2.support.design.widget.RxNavigationView;
import com.nrs.nsnik.notes.BuildConfig;
import com.nrs.nsnik.notes.MyApplication;
import com.nrs.nsnik.notes.R;
import com.nrs.nsnik.notes.util.idlingResource.SimpleIdlingResource;
import com.nrs.nsnik.notes.view.fragments.AboutFragment;
import com.nrs.nsnik.notes.view.fragments.HomeFragment;
import com.squareup.leakcanary.RefWatcher;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.disposables.CompositeDisposable;


public class MainActivity extends AppCompatActivity {

    private static final String[] mFragTags = {"home", "starred", "vault", "about"};
    @Nullable
    @BindView(R.id.mainDrawerLayout)
    DrawerLayout mDrawerLayout;
    @Nullable
    @BindView(R.id.mainNavigationView)
    NavigationView mNavigationView;
    @Nullable
    @BindView(R.id.mainToolbar)
    Toolbar mMainToolbar;
    @Nullable
    @BindView(R.id.mainToolbarText)
    TextView mToolbarText;
    private String mCurrentFragment;
    private boolean mIsClicked;
    private CompositeDisposable mCompositeDisposable;

    @Nullable
    private SimpleIdlingResource mIdlingResource;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTheme(R.style.transparentStatusBar);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        initialize(savedInstanceState);
        initializeDrawer();
        getIdlingResource();
    }

    @VisibleForTesting
    @NonNull
    public IdlingResource getIdlingResource() {
        if (mIdlingResource == null) {
            mIdlingResource = new SimpleIdlingResource();
        }
        return mIdlingResource;
    }

    private void initialize(@Nullable Bundle savedInstanceState) {

        if (savedInstanceState == null && mToolbarText != null) {
            setSupportActionBar(mMainToolbar);
            if (getSupportActionBar() != null) {
                getSupportActionBar().setDisplayShowTitleEnabled(false);
            }

            mCurrentFragment = mFragTags[0];

            getSupportFragmentManager().beginTransaction().replace(R.id.mainContainer, new HomeFragment(), mCurrentFragment).commit();
            mToolbarText.setText(getResources().getString(R.string.app_name));
        }
        mCompositeDisposable = new CompositeDisposable();
    }

    /**
     * initialize the navigation drawer by attaching the
     * listeners and item click listeners
     */
    private void initializeDrawer() {
        ActionBarDrawerToggle actionBarDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout, mMainToolbar, R.string.drawerOpen, R.string.drawerClose) {
            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
            }

            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
                if (mIsClicked) {
                    mIsClicked = false;
                    performChange();
                }
            }
        };
        if (mDrawerLayout != null) {
            mDrawerLayout.addDrawerListener(actionBarDrawerToggle);
        }
        actionBarDrawerToggle.syncState();
        if (mNavigationView != null) {
            mNavigationView.getMenu().getItem(0).setChecked(true);
        }
        if (mNavigationView != null) {
            mCompositeDisposable.add(RxNavigationView.itemSelections(mNavigationView).subscribe(menuItem -> {
                if (mIdlingResource != null) {
                    mIdlingResource.setIdleState(false);
                }
                switch (menuItem.getItemId()) {
                    case R.id.navItem1:
                        if (mCurrentFragment != null && !mCurrentFragment.equalsIgnoreCase(mFragTags[0])) {
                            mCurrentFragment = mFragTags[0];
                            mIsClicked = true;
                        }
                        break;
                    case R.id.navItem2:
                        Toast.makeText(MainActivity.this, "To-Do", Toast.LENGTH_LONG).show();
                        break;
                    case R.id.navItem3:
                        Toast.makeText(MainActivity.this, "To-Do", Toast.LENGTH_LONG).show();
                        break;
                    case R.id.navItem4:
                        Toast.makeText(MainActivity.this, "To-Do", Toast.LENGTH_LONG).show();
                        break;
                    case R.id.navItem5:
                        if (mCurrentFragment != null && !mCurrentFragment.equalsIgnoreCase(mFragTags[3])) {
                            mCurrentFragment = mFragTags[3];
                            mIsClicked = true;
                        }
                        break;
                }
                mDrawerLayout.closeDrawers();
            }));
        }
    }

    private void performChange() {
        if (mToolbarText != null) {
            if (mCurrentFragment.equalsIgnoreCase(mFragTags[0])) {
                replaceFragment(new HomeFragment(), mCurrentFragment);
                mToolbarText.setText(getResources().getString(R.string.app_name));
            } else if (mCurrentFragment.equalsIgnoreCase(mFragTags[1])) {
                Toast.makeText(MainActivity.this, "To-Do", Toast.LENGTH_LONG).show();
            } else if (mCurrentFragment.equalsIgnoreCase(mFragTags[2])) {
                Toast.makeText(MainActivity.this, "To-Do", Toast.LENGTH_LONG).show();
            } else if (mCurrentFragment.equalsIgnoreCase(mFragTags[3])) {
                replaceFragment(new AboutFragment(), mCurrentFragment);
                mToolbarText.setText(getResources().getString(R.string.navItem5));
            }
        }
        if (mIdlingResource != null) {
            mIdlingResource.setIdleState(true);
        }
    }

    /**
     * @param fragment the new fragment that will replace the old fragment
     * @param tag      the tag for the new fragment
     */
    public void replaceFragment(Fragment fragment, String tag) {
        getSupportFragmentManager()
                .beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.mainContainer, fragment, tag)
                .commit();
    }

    private void cleanUp() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable.clear();
            mCompositeDisposable.dispose();
        }
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menuMainSearch:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cleanUp();
        if (BuildConfig.DEBUG) {
            RefWatcher refWatcher = MyApplication.getRefWatcher(this);
            refWatcher.watch(this);
        }
    }
}

File: app/src/main/java/com/nrs/nsnik/notes/view/MainActivity.kt
/*
 *     Notes  Copyright (C) 2018  Nikhil Soni
 *     This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
 *     This is free software, and you are welcome to redistribute it
 *     under certain conditions; type `show c' for details.
 *
 * The hypothetical commands `show w' and `show c' should show the appropriate
 * parts of the General Public License.  Of course, your program's commands
 * might be different; for a GUI interface, you would use an "about box".
 *
 *   You should also get your employer (if you work as a programmer) or school,
 * if any, to sign a "copyright disclaimer" for the program, if necessary.
 * For more information on this, and how to apply and follow the GNU GPL, see
 * <http://www.gnu.org/licenses/>.
 *
 *   The GNU General Public License does not permit incorporating your program
 * into proprietary programs.  If your program is a subroutine library, you
 * may consider it more useful to permit linking proprietary applications with
 * the library.  If this is what you want to do, use the GNU Lesser General
 * Public License instead of this License.  But first, please read
 * <http://www.gnu.org/philosophy/why-not-lgpl.html>.
 */

package com.nrs.nsnik.notes.view

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.view.MenuItem
import android.view.View
import androidx.annotation.NonNull
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import androidx.test.espresso.IdlingResource
import com.nrs.nsnik.notes.BuildConfig
import com.nrs.nsnik.notes.MyApplication
import com.nrs.nsnik.notes.R
import com.nrs.nsnik.notes.util.idlingResource.SimpleIdlingResource
import io.branch.referral.Branch
import kotlinx.android.synthetic.main.activity_main.*
import org.json.JSONException
import timber.log.Timber


class MainActivity : AppCompatActivity() {

    private var mIdlingResource: SimpleIdlingResource? = null
    private var isOnListFragment: Boolean = true

    companion object {
        private const val SearchIntentAction: String = "com.nrs.nsnik.notes.StartSearch"
        private const val NewNoteIntentAction: String = "com.nrs.nsnik.notes.OpenNewNotesFragment"
    }

    override fun onStart() {
        super.onStart()
        Branch.getInstance().initSession({ referringParams, error ->
            Timber.d(if (error == null) referringParams.toString() else error.message)
        }, this.intent.data, this)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initialize()
        handleIntent()
    }

//override fun onSupportNavigateUp() = findNavController(R.id.mainNavHost).navigateUp()

    override fun onResume() {
        super.onResume()
        if (Branch.isAutoDeepLinkLaunch(this)) {
            try {
                val autoDeeplinkedValue = Branch.getInstance().latestReferringParams.getString("noteTitle")
                Timber.d("Launched by Branch on auto deep linking!\n\n$autoDeeplinkedValue")
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        } else {
            Timber.d("Launched by normal application flow")
        }
    }

    private fun initialize() {
        setSupportActionBar(mainToolbar)
        supportActionBar?.apply {
            //            setDisplayHomeAsUpEnabled(true)
//            setHomeAsUpIndicator(R.drawable.ic_baseline_menu_24px)
        }

        val controller = findNavController(R.id.mainNavHost)


        controller.addOnDestinationChangedListener { navController, destination, arguments ->
            isOnListFragment = destination.id == R.id.navItemNotes
            mainToolbar.visibility = if (destination.id == R.id.introFragment) View.GONE else View.VISIBLE
            mainDrawerLayout.setDrawerLockMode(if (destination.id == R.id.navItemNotes) DrawerLayout.LOCK_MODE_UNLOCKED else DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
            mainNavigationView.setCheckedItem(
                    when (destination.id) {
                        R.id.navItemNotes -> R.id.navItemNotes
                        R.id.navItemSettings -> R.id.navItemSettings
                        R.id.navItemAbout -> R.id.navItemAbout
                        else -> R.id.navItemNotes
                    }
            )
        }

        setupActionBarWithNavController(this, controller, mainDrawerLayout)

        AppBarConfiguration(controller.graph, mainDrawerLayout)

        mainNavigationView.setupWithNavController(controller)

        mainToolbar.setupWithNavController(controller)

        mainNavigationView.setNavigationItemSelectedListener { menuItem ->
            when (menuItem.itemId) {
                R.id.navItemNotes -> {
                    mainDrawerLayout.closeDrawer(GravityCompat.START)
                    return@setNavigationItemSelectedListener true
                }
                R.id.navItemSettings -> {
                    mainDrawerLayout.closeDrawer(GravityCompat.START)
                    findNavController(R.id.mainNavHost).navigate(R.id.navItemSettings)
                    return@setNavigationItemSelectedListener true
                }
                R.id.navItemAbout -> {
                    mainDrawerLayout.closeDrawer(GravityCompat.START)
                    findNavController(R.id.mainNavHost).navigate(R.id.navItemAbout)
                    return@setNavigationItemSelectedListener true
                }
                else -> return@setNavigationItemSelectedListener false
            }
        }
    }

    private fun handleIntent() {
        val intent = intent
        val action = intent.action
        val type = intent.type

        when (action) {
            Intent.ACTION_SEND -> {
                if (type != null)
                    if ("text/plain" == type) handleSendText(intent) else if (type.startsWith("image/")) handleSendImage(intent)
            }
            Intent.ACTION_SEND_MULTIPLE -> {
                if (type != null)
                    if (type.startsWith("image/")) handleSendMultipleImages(intent)
            }
            NewNoteIntentAction -> findNavController(R.id.mainNavHost).navigate(R.id.newNoteFragment)
            SearchIntentAction -> findNavController(R.id.mainNavHost).navigate(R.id.searchFragment)
            else -> {
                // Handle other intents, such as being started from the home screen
            }
        }
    }

    private fun handleSendText(intent: Intent) {
        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (sharedText != null) {
            val bundle = Bundle()
            bundle.putString(resources.getString(R.string.bundleReceiveIntentText), sharedText)
            findNavController(R.id.mainNavHost).navigate(R.id.newNoteFragment, bundle)
        }
    }

    private fun handleSendImage(intent: Intent) {
        val imageUri = intent.getParcelableExtra<Parcelable>(Intent.EXTRA_STREAM) as Uri
        val bundle = Bundle()
        bundle.putParcelable(resources.getString(R.string.bundleReceiveIntentImage), imageUri)
        findNavController(R.id.mainNavHost).navigate(R.id.newNoteFragment, bundle)
    }

    private fun handleSendMultipleImages(intent: Intent) {
        val imageUris = intent.getParcelableArrayListExtra<Parcelable>(Intent.EXTRA_STREAM)
        if (imageUris != null) {
            val bundle = Bundle()
            bundle.putParcelableArrayList(resources.getString(R.string.bundleReceiveIntentImageList), imageUris)
            findNavController(R.id.mainNavHost).navigate(R.id.newNoteFragment, bundle)
        }
    }

    override fun onBackPressed() {
        if (mainDrawerLayout.isDrawerOpen(GravityCompat.START)) mainDrawerLayout.closeDrawer(GravityCompat.START)
        else super.onBackPressed()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> if (isOnListFragment) mainDrawerLayout.openDrawer(GravityCompat.START)
        }
        return super.onOptionsItemSelected(item)
    }

    @VisibleForTesting
    @NonNull
    fun getIdlingResource(): IdlingResource {
        if (mIdlingResource == null) mIdlingResource = SimpleIdlingResource()
        return mIdlingResource as IdlingResource
    }

    override fun onDestroy() {
        super.onDestroy()
        if (BuildConfig.DEBUG) MyApplication.getRefWatcher(this)?.watch(this)
    }
}